core/src/dsnp/api_types.rs

```diff
@@ -0,0 +1,96 @@
+#![allow(dead_code)] // todo: remove after usage
+use crate::dsnp::{dsnp_types::DsnpId, encryption::EncryptionBehavior};
+
+/// Raw page of Graph (or Key) data
+pub type Page = Vec<u8>;
+
+pub struct KeyPair<E: EncryptionBehavior> {
+	/// Public key
+	pub public_key: E::EncryptionInput,
+
+	/// Private key
+	pub private_key: E::DecryptionInput,
+}
+
+pub type PublicKey<E> = <E as EncryptionBehavior>::EncryptionInput;
+
+pub enum PrivacyType {
+	Public,
+	Private,
+}
+
+pub enum ConnectionType {
+	Follow(PrivacyType),
+	Friendship(PrivacyType),
+}
+
+pub trait Config {
+	/// Graph page id
+	type PageId;
+	/// Schema ID
+	type SchemaId;
+
+	const PUBLIC_FOLLOW_SCHEMAID: Self::SchemaId;
+	const PRIVATE_FOLLOW_SCHEMAID: Self::SchemaId;
+	const PUBLIC_FRIEND_SCHEMAID: Self::SchemaId;
+	const PRIVATE_FRIEND_SCHEMAID: Self::SchemaId;
+
+	fn schema_for_connection_type(connection_type: ConnectionType) -> Self::SchemaId {
+		match connection_type {
+			ConnectionType::Follow(PrivacyType::Public) => Self::PUBLIC_FOLLOW_SCHEMAID,
+			ConnectionType::Follow(PrivacyType::Private) => Self::PRIVATE_FOLLOW_SCHEMAID,
+			ConnectionType::Friendship(PrivacyType::Public) => Self::PUBLIC_FRIEND_SCHEMAID,
+			ConnectionType::Friendship(PrivacyType::Private) => Self::PRIVATE_FRIEND_SCHEMAID,
+		}
+	}
+}
+
+pub struct Import<E: EncryptionBehavior> {
+	pub dsnp_user_id: DsnpId,
+	pub keys: Vec<KeyPair<E>>,
+	pub pages: Vec<Page>,
+}
+
+pub struct Connection {
+	pub dsnp_user_id: DsnpId,
+	pub connection_type: ConnectionType,
+}
+
+pub struct DsnpKeys<E: EncryptionBehavior> {
+	pub dsnp_user_id: DsnpId,
+	pub keys: Vec<PublicKey<E>>,
+}
+
+pub enum Action<E: EncryptionBehavior> {
+	Connect {
+		owner_dsnp_user_id: DsnpId,
+		connection: Connection,
+		connection_key: Option<PublicKey<E>>, // included only if PRId calculation is required
+	},
+	Disconnect {
+		owner_dsnp_user_id: DsnpId,
+		connection: Connection,
+	},
+}
+
+pub enum Update<C: Config> {
```

Comment:
I'm guessing using generic types like this would be really hard to make it compatible with C.
My assumption is that associated types makes sense if you might have different types for different  implementations but I don't think we will have different implementations for `PageId` and `SchemaId`